test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test;
+
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+import static org.elasticsearch.test.ESTestCase.assertEquals;
+
+/**
+ * Enum serialization via {@link StreamOutput#writeEnum} and {@link StreamInput#readEnum(Class)} uses the enum value's ordinal on the wire.
+ * Reordering the values in an enum, or adding a new value, will change the ordinals and is therefore a wire protocol change, but it's easy
+ * to miss this fact in the context of a larger commit. To protect against this trap, any enums that we send over the wire should have a
+ * test that uses {@link #assertEnumSerialization} to assert a fixed mapping between ordinals and values. That way, a change to the ordinals
+ * will require a test change, and thus some thought about BwC.
+ */
+public class EnumSerializationTestUtils {
+    private EnumSerializationTestUtils() {/* no instances */}
+
+    /**
+     * Assert that the enum constants of the given class are exactly the ones passed in explicitly as arguments, which fixes its wire
+     * protocol when using {@link StreamOutput#writeEnum} and {@link StreamInput#readEnum(Class)}.
+     *
+     * @param value0 The enum constant with ordinal {@code 0}, passed as a separate argument to avoid prevent callers just lazily using
+     *               {@code EnumClass.values()} to pass the values of the enum, which would negate the point of this test.
```

Comment:
Would `EnumClass.values()` return the same array as `clazz.getEnumConstants()` in such case?
If so we could protect against this by `assertNotSame(enumConstants, values)`

Reply:
`EnumClass.values()` doesn't seem to work with generics.

Reply:
Nope, I do not suggest to create it from generics, but verify that the provided array is not created by it.

Reply:
I don't see how but maybe I'm missing something. Could you check out this branch and try it out in IntelliJ, then share a patch with what you mean?

Reply:
```
Index: server/src/test/java/org/elasticsearch/transport/RemoteConnectionStrategyTests.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/server/src/test/java/org/elasticsearch/transport/RemoteConnectionStrategyTests.java b/server/src/test/java/org/elasticsearch/transport/RemoteConnectionStrategyTests.java
--- a/server/src/test/java/org/elasticsearch/transport/RemoteConnectionStrategyTests.java	(revision bcf59ae880a669682bf55f503eff2ba521365789)
+++ b/server/src/test/java/org/elasticsearch/transport/RemoteConnectionStrategyTests.java	(date 1721389890801)
@@ -165,6 +165,13 @@
         );
     }
 
+    public void testConnectionStrategySerialization2() {
+        EnumSerializationTestUtils.assertEnumSerialization(
+            RemoteConnectionStrategy.ConnectionStrategy.class,
+            RemoteConnectionStrategy.ConnectionStrategy.values()
+        );
+    }
+
     private static class FakeConnectionStrategy extends RemoteConnectionStrategy {
 
         private final ConnectionStrategy strategy;
Index: test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java b/test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java
--- a/test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java	(revision bcf59ae880a669682bf55f503eff2ba521365789)
+++ b/test/framework/src/main/java/org/elasticsearch/test/EnumSerializationTestUtils.java	(date 1721390017670)
@@ -15,6 +15,7 @@
 import java.io.IOException;
 
 import static org.elasticsearch.test.ESTestCase.assertEquals;
+import static org.junit.Assert.assertNotSame;
 
 /**
  * Enum serialization via {@link StreamOutput#writeEnum} and {@link StreamInput#readEnum(Class)} uses the enum value's ordinal on the wire.
@@ -29,17 +30,16 @@
     /**
      * Assert that the enum constants of the given class are exactly the ones passed in explicitly as arguments, which fixes its wire
      * protocol when using {@link StreamOutput#writeEnum} and {@link StreamInput#readEnum(Class)}.
-     *
-     * @param value0 The enum constant with ordinal {@code 0}, passed as a separate argument to avoid prevent callers just lazily using
-     *               {@code EnumClass.values()} to pass the values of the enum, which would negate the point of this test.
+
      * @param values The remaining enum constants, in ordinal order.
      */
     @SafeVarargs
-    public static <E extends Enum<E>> void assertEnumSerialization(Class<E> clazz, E value0, E... values) {
+    public static <E extends Enum<E>> void assertEnumSerialization(Class<E> clazz, E... values) {
         final var enumConstants = clazz.getEnumConstants();
-        assertEquals(clazz.getCanonicalName(), enumConstants.length, values.length + 1);
-        for (var ordinal = 0; ordinal < values.length + 1; ordinal++) {
-            final var enumValue = ordinal == 0 ? value0 : values[ordinal - 1];
+        assertNotSame("Should list values explicitly rather than via .values()", enumConstants, values);
+        assertEquals(clazz.getCanonicalName(), enumConstants.length, values.length);
+        for (var ordinal = 0; ordinal < values.length; ordinal++) {
+            final var enumValue = values[ordinal];
             final var description = clazz.getCanonicalName() + "[" + ordinal + "]";
             assertEquals(description, enumConstants[ordinal], enumValue);
             try (var out = new BytesStreamOutput(1)) {
```

I was hoping this would work, but apparently `getEnumConstants` clones array every time, so it is noever the same as `values()`.

Reply:
Oh right I see, I thought you meant to call `values` in the test util. Yes, not to worry, we want to be explicit about the enum values anyway.